docs/io/ascii/ecsv.rst

```diff
@@ -0,0 +1,471 @@
+----------------------
+
+ECSV supports storing multidimensional columns is when the length of each array
+element may vary. This data structure is supported in the `FITS standard
+<https://fits.gsfc.nasa.gov/fits_standard.html>`_. While ``numpy`` does not
+natively support variable-length arrays, it is possible to represent such a
+structure using an object-type array of typed ``np.ndarray`` objects. This is how
+the ``astropy`` FITS reader outputs a variable-length array.
+
+This capability is added in astropy 4.3 and ECSV version 1.0.
+
+Most commonly variable-length arrays have a 1-d array in each cell of the
+column. You might a column with 1-d ``np.ndarray`` cells having lengths of 2, 5,
+and 3 respectively.
+
+The ECSV standard and ``astropy`` also supports arbitrary N-d arrays in each
+cell, where all dimensions except the last one must match. For instance you
+could have a column with ``np.ndarray`` cells having shapes of ``(4,4,2)``,
+``(4,4,5)``, and ``(4,4,3)`` respectively.
+
+..
+  EXAMPLE START
+  Using ECSV Format to Write Astropy Tables with Variable-Length Arrays
+
+The example below shows writing a variable-length 1-d array to ECSV. Notice the
+new ECSV column attribute ``subtype: 'int64[null]'``. The ``[null]`` indicates a
+variable length for the one dimension. If we had been writing the N-d example
+above the subtype would have been ``int64[4,4,null]``.
+
+.. doctest-skip::
+
+  >>> t = Table()
+  >>> t['a'] = np.empty(3, dtype=object)
+  >>> t['a'] = [np.array([1, 2], dtype=np.int64),
+  ...           np.array([3, 4, 5], dtype=np.int64),
+  ...           np.array([6, 7, 8, 9], dtype=np.int64)]
+  >>> ascii.write(t, format='ecsv')
+  # %ECSV 1.0
+  # ---
+  # datatype:
+  # - {name: a, datatype: string, subtype: 'int64[null]'}
+  # schema: astropy-2.0
+  a
+  [1,2]
+  [3,4,5]
+  [6,7,8,9]
+
+..
+  EXAMPLE END
+
+Object arrays
+-------------
+
+ECSV can store object-type columns with simple Python objects consisting of
+``dict``, ``list``, ``str``, ``int``, ``float``, ``bool`` and ``None`` elements.
+More precisely, any object that can be serialized to `JSON
+<https://www.json.org/>`__ using the standard library `json
+<https://docs.python.org/3/library/json.html>`__ package is supported.
+
+..
+  EXAMPLE START
+  Using ECSV Format to Write Astropy Tables with Object Arrays
+
+The example below shows writing an object array to ECSV. Because JSON requires
+a double-quote around strings, and because ECSV requires ``""`` to represent
+a double-quote within a string, one tends to get double-double quotes in this
+representation.
+
+.. doctest-skip::
+
+  >>> t = Table()
+  >>> t['a'] = np.array([{'a': 1},
+  ...                    {'b': [2.5, None]},
+  ...                    True], dtype=object)
+  >>> ascii.write(t, format='ecsv')
+  # %ECSV 1.0
+  # ---
+  # datatype:
+  # - {name: a, datatype: string, subtype: object}
+  # schema: astropy-2.0
+  a
+  "{""a"":1}"
+  "{""b"":[2.5,null]}"
+  true
+
+..
+  EXAMPLE END
```

Comment:
Really nice addition! One question: what makes it necessary to skip most of the writing tests?

Reply:
I was getting RTD fails if I don't skip them.